csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -312,16 +316,33 @@ void propagateDIDTransform(
     TensorView* tv,
     const std::unordered_set<ParallelType>& selected_parallel_types,
     PropagateDirection direction) {
-  tv->setDeviceMesh(ref->getDeviceMesh());
-
   std::unordered_set<IterDomain*> device_ids;
   const std::unordered_map<IterDomain*, IterDomain*> ref2target =
       getRef2TargetMap(ref, tv, direction);
 
+  const auto& ref_mesh = ref->getDeviceMesh();
+
   for (IterDomain* device_id : ref->getLoopDomain()) {
     if (selected_parallel_types.count(device_id->getParallelType()) == 0) {
       continue;
     }
+
+    // If the target has a device mesh, the device id can only be propagated
+    // iff:
+    // 1. The device id parallel type is present in the device mesh.
+    // 2. The number of devices corresponding to the parallel type is same
+    // between the reference and target.
```

Comment:
We may need further verification, such as checking the actual device indices. Leaving this to future work, since I am not quite certain in the absence of motivating cases.